### PR TITLE
MP-23, MP-24, MP-47 - Searching and pagination mechanisms. Displaying data from Firebase.

### PR DESCRIPTION
**Pull request by mfurmanit**

### CODE REVIEW

File: src/react-table-config.d.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  UseRowStateState,
+  UseSortByColumnProps,
+  UseSortByHooks,
+  UseSortByInstanceProps,
+  UseSortByOptions,
+  UseSortByState,
+  UseTableInstanceProps,
+} from 'react-table';
+
+declare module 'react-table' {
```
Comment (mfurmanit) [inline src/react-table-config.d.ts:47]: Used / made by referring to this [description](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-table/Readme.md).

**Review by kacpergadula — APPROVED** [review]

File: src/api/firebase/firebase.types.ts [hunk 2]
```diff
@@ -1,9 +1,8 @@
-/* eslint-disable max-len */
 import firebase from 'api/firebase/firebase.api';
 
-export type Firestore = {
-  Doc: firebase.firestore.QueryDocumentSnapshot<firebase.firestore.DocumentData>;
-};
+export type Snapshot = firebase.firestore.QuerySnapshot<firebase.firestore.DocumentData>;
```
Comment (dominikmlynarczyk) [inline src/api/firebase/firebase.types.ts:3]: Why did you remove the object type in favor of separately exported types? Generally this is not a problem, but as we use Firebase services such as Storage, so as not to confuse e.g. documents from these two services the `Firebase.` prefix might be useful.

BTW (correct me if I'm wrong) - The `QueryDocumentSnapshot` type is used when the GET request is made, and the `QuerySnapshot` type is used when the `onSnaphot()` method is used or is it the other way around?

File: src/api/firebase/firestore/firestore-actions.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { BaseObject } from 'model/base-object';
+import { CollectionReference, Documents, Snapshot } from '../firebase.types';
+import { firestore } from '../firebase.api';
+
+export const all = async <T extends BaseObject>(collection: string): Promise<T[]> => {
```
Comment (dominikmlynarczyk) [inline src/api/firebase/firestore/firestore-actions.ts]: Correct me if I'm wrong, but are you making an `all` function that takes an individual collection? What is the purpose?
I don't think `all` is a descriptive name, what is the purpose of this function?

File: src/others/theme.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import { extendTheme } from '@chakra-ui/react';
+
+export const theme = extendTheme({
```
Comment (dominikmlynarczyk) [inline src/others/theme.ts:3]: What is the purpose? I'm not convinced that we have a specific reason to do custom breakpoints at this stage of the application. [Responsive styles at Chakra doc](https://chakra-ui.com/docs/features/responsive-styles)

File: src/components/Table/AdvancedTable.tsx [hunk 5]
```diff
@@ -1,33 +1,90 @@
...
+
+  const onSearch = (value: string): void => {
+    if (isEmpty(value) || isEmpty(props.filterColumns)) setDisplayedData(data);
+    else setDisplayedData(data.filter((row) => filterRow(row, value)));
+  };
+
+  const filterRow = (row: T, value: string): boolean => {
+    let show: boolean = false;
+    if (props.filterColumns !== undefined) {
+      for (let i = 0; i < props.filterColumns.length; i += 1) {
```
Comment (dominikmlynarczyk) [inline src/components/Table/AdvancedTable.tsx:50]: Wouldn't it be better if you wrote that imperative code more declaratively, e.g. using the array filter method?

File: src/components/Table/AdvancedTable.tsx [hunk 6]
```diff
@@ -1,33 +1,90 @@
...
+      });
+  }, [props.collection]);
+
+  const onSearch = (value: string): void => {
+    if (isEmpty(value) || isEmpty(props.filterColumns)) setDisplayedData(data);
+    else setDisplayedData(data.filter((row) => filterRow(row, value)));
+  };
+
+  const filterRow = (row: T, value: string): boolean => {
+    let show: boolean = false;
```
Comment (dominikmlynarczyk) [inline src/components/Table/AdvancedTable.tsx:48]: Use proper naming convention for variables which are storing boolean values. What is `show`?
Could you change that to `isVisible` or something that is more descriptive?

File: src/components/Table/AdvancedTable.tsx [hunk 7]
```diff
@@ -1,33 +1,90 @@
...
+        setDisplayedData(categories);
+      })
+      .catch((error) => {
+        throw new Error(`Could not fetch ${props.collection} collection!. Error: ${error.message}`);
+      });
+  }, [props.collection]);
+
+  const onSearch = (value: string): void => {
+    if (isEmpty(value) || isEmpty(props.filterColumns)) setDisplayedData(data);
+    else setDisplayedData(data.filter((row) => filterRow(row, value)));
```
Comment (dominikmlynarczyk) [inline src/components/Table/AdvancedTable.tsx:44]: I see that you are doing another loop inside filterRow which is a O(2) complexity... Is it necessary?

File: src/components/Table/Body.tsx [hunk 8]
```diff
@@ -1,19 +1,20 @@
...
-type BodyProps<T extends object> = {
+type BodyProps<T extends BaseObject> = {
   bodyProps: TableBodyProps;
-  rows: Array<Row<T>>;
+  page: Array<Row<T>>;
   prepareRow: (row: Row<T>) => void;
 };
 
-function Body<T extends object>(props: BodyProps<T>) {
+function Body<T extends BaseObject>(props: BodyProps<T>) {
```
Comment (dominikmlynarczyk) [inline src/components/Table/Body.tsx:14]: Destructure props please.

File: src/components/Table/Paginator.tsx [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+      m={{ base: 0, lg: 4 }}
+      mt={{ base: 4 }}
+      flexWrap={{ base: 'wrap', lg: 'nowrap' }}
+      flexDirection={{ base: 'row' }}
+    >
+      <Flex order={1}>
+        <Tooltip label="Pierwsza strona">
+          <IconButton
+            aria-label="Pierwsza strona"
+            onClick={() => props.gotoPage(0)}
```
Comment (dominikmlynarczyk) [inline src/components/Table/Paginator.tsx:31]: Destructure the props as much as possible everywhere please.

**Review by dominikmlynarczyk — CHANGES_REQUESTED** [review]
Thank you for your PR.
Mostly just discuss the changes together, in only few cases just correct code according to the review.

File: src/api/firebase/firestore/firestore-actions.ts (hunk 3 above)
Comment (mfurmanit) [reply src/api/firebase/firestore/firestore-actions.ts]: The purpose of `all` method is to fetch all documents from collection specified in parameter.
I can name the method as you wish - `getAll`, `fetchAll` etc. Which do you prefer?

This file was intended to be something like `global service`, but so far there isn't much logic.
I want to store here methods like `create`, `update` and `delete` that will be shared by every table / details page.

File: src/others/theme.ts (hunk 4 above)
Comment (mfurmanit) [reply src/others/theme.ts:3]: Actually there are no breakpoints in this file.
I wanted to override default `min-width` property for `body` tag - nothing else.

File: src/api/firebase/firebase.types.ts (hunk 2 above)
Comment (mfurmanit) [reply src/api/firebase/firebase.types.ts:3]: I have never seen defining object types like this before.
For me types should be declared separately and that's how you see it most of the time in TypeScript projects.

`QuerySnapshot` is returned after firing `get` method on `CollectionReference`.
`QueryDocumentSnapshot` type is some kind of wrapper for retrieved firebase `docs`.
image

Comment (mfurmanit) [reply src/api/firebase/firebase.types.ts:3]: We discussed at the meeting on 23.04.2021 that we're leaving types this way for now.

File: src/api/firebase/firestore/firestore-actions.ts (hunk 3 above)
Comment (mfurmanit) [reply src/api/firebase/firestore/firestore-actions.ts]: We discussed at the meeting on 23.04.2021 that we're leaving generic types this way for now.
We also came to the conclusion that `getAll` would be a better name for that function. I changed that.

File: src/components/Table/AdvancedTable.tsx (hunk 5 above)
Comment (mfurmanit) [reply src/components/Table/AdvancedTable.tsx:50]: We discussed at the meeting on 23.04.2021 that I'm leaving it this way for now.

File: src/components/Table/AdvancedTable.tsx (hunk 6 above)
Comment (mfurmanit) [reply src/components/Table/AdvancedTable.tsx:48]: We discussed at the meeting on 23.04.2021 that I'm leaving it this way for now - it's fine.

File: src/components/Table/AdvancedTable.tsx (hunk 7 above)
Comment (mfurmanit) [reply src/components/Table/AdvancedTable.tsx:44]: We discussed at the meeting on 23.04.2021 that I'm leaving it this way for now. If we find a better solution, we will change it in the future.

File: src/components/Table/Body.tsx (hunk 8 above)
Comment (mfurmanit) [reply src/components/Table/Body.tsx:14]: We came to the conclusion (at the meeting on 23.04.2021) that destructing isn't required and necessary due to code duplication (types are declared near `props` in those components).

File: src/components/Table/Paginator.tsx (hunk 9 above)
Comment (mfurmanit) [reply src/components/Table/Paginator.tsx:31]: We came to the conclusion (at the meeting on 23.04.2021) that destructing isn't required and necessary due to code duplication (types are declared near `props` in those components).

File: src/others/theme.ts (hunk 4 above)
Comment (mfurmanit) [reply src/others/theme.ts:3]: We discussed at the meeting on 23.04.2021 that I'm leaving it this way for now. 
We also came to the conclusion that 320px would be better value. I changed that.

**Pull request merged**